Assign error codes to all of our messages

nguerrera opened the pull request:
Fix #2065 

This adds "NETSDK1234" error codes to each of our error/warning/log messages and sends them to msbuild as such.

We need to add a new message with a code in order for project system to light up an info bar. This puts the infrastructure in place for that.

Individual commits may be easier to review.

The most interesting commits are:
* 9596e8d - Shows the pattern for adding a new error code to resx. From there we can call Log.LogXxx as before.
* d3ffdf8 - Implements the plumbing of the codes from resx to msbuild.

nguerrera committed:
Remove unused strings

nguerrera committed:
Rename resource string that looks like NuGet error code, but isn't

nguerrera committed:
Assign error codes to strings

nguerrera committed:
Update XLF

nguerrera committed:
Use latest C# language version

nguerrera committed:
Rework task logging to honor error codes in resx

nguerrera committed:
Remove common source files from solution items

The experience of having them show up a 3rd time in solution
explorer is not good. Plus, having .cs files listed in sln
defeats the benefits of globbing and it is easy for them
to get out of date.

nguerrera committed:
Add unit tests for new task logging mechanism

dsplaisted commented on src/Tasks/Common/Resources/Strings.resx:
```diff
@@ -118,202 +118,267 @@
   </resheader>
   <data name="AtLeastOneTargetFrameworkMustBeSpecified" xml:space="preserve">
-    <value>At least one possible target framework must be specified.</value>
+    <value>NETSDK1001: At least one possible target framework must be specified.</value>
```
Let's add tests that validate that we don't duplicate error codes over time.

To ensure that this still holds even for deleted messages, we discussed having a list of deleted message codes, and ensuring that there are no duplicates between the current and deleted message codes, and that the current and deleted message codes are contiguous.

nguerrera replied:
Done.

dsplaisted commented on src/Tasks/Common/Logger.cs:
```diff
@@ -0,0 +1,123 @@
+    ///     Log.LogError(Strings.SomethingIsWrong);
+    ///
+    ///   Strings.resx:
+    ///     | SomethingIsWrong | NETSDK1234: Something is wrong. |
```
The pipes didn't make it clear to me that this wasn't just one long string.  Perhaps something like this would be clearer:

```
  Strings.resx:
    Resource Name: SomethingIsWrong
    Resource Value: NETSDK1234: Something is wrong.
```

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/LockFileCache.cs:
```diff
@@ -107,28 +107,26 @@ private LockFile LoadLockFile(string path)
+                    new Message(
+                        level: message.Level == LogLevel.Warning ? MessageLevel.Warning : MessageLevel.NormalImportance,
+                        code: message.Code == NuGetLogCode.Undefined ? default : message.Code.ToString(),
+                        text: message.Message));
```
Plumb the file through here from the NuGet message.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenALogger.cs:
```diff
@@ -0,0 +1,104 @@
+        {
+            var logger = new TestLogger();
+
+            logger.LogError("No error code");
```
Won't this fail in Debug?  I think it needs to be moved into the `#else` clause, or possibly delete it, bring `logWithoutErrorCode` out of the `#if` clause, and invoke it in the `#else` clause.

nguerrera replied:
Oops.

dsplaisted reviewed (approved):
Looks good, please add tests to ensure we don't duplicate message codes as discussed.

nguerrera committed:
Add test to ensure error codes are used correctly

1. All messages need error codes in the correct format
2. Error codes cannot be duplicated
3. Error codes cannot be skipped

nguerrera commented:
> please add tests to ensure we don't duplicate message codes as discussed.

@dsplaisted Done in b038a28

The other commits since review are fixups to be rebased away that 1) fix bugs found by the test (yay!) 2) adress other minor feedback.
